File: examples/signature.java
package imgproxytest;

import org.junit.jupiter.api.Test;

import java.util.Base64;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import static org.junit.jupiter.api.Assertions.*;

public class ImgProxy{

    @Test
    public void testWithJavaHmacApacheBase64ImgProxyTest() throws Exception {
        byte[] key = hexStringToByteArray("943b421c9eb07c830af81030552c86009268de4e532ba2ee2eab8247c6da0881");
        byte[] salt = hexStringToByteArray("520f986b998545b4785e0defbc4f3c1203f22de2374a3d53cb7a7fe9fea309c5");

        String path = "/rs:fit:300:300/plain/http://img.example.com/pretty/image.jpg";

        String pathWithHash = signPath(key, salt, path);

        assertEquals("/m3k5QADfcKPDj-SDI2AIogZbC3FlAXszuwhtWXYqavc/rs:fit:300:300/plain/http://img.example.com/pretty/image.jp", pathWithHash);
    }

    public static String signPath(byte[] key, byte[] salt, String path) throws Exception {
        final String HMACSHA256 = "HmacSHA256";

        Mac sha256HMAC = Mac.getInstance(HMACSHA256);
        SecretKeySpec secretKey = new SecretKeySpec(key, HMACSHA256);
        sha256HMAC.init(secretKey);
        sha256HMAC.update(salt);

        String hash = Base64.getUrlEncoder().withoutPadding().encodeToString(sha256HMAC.doFinal(path.getBytes()));

        return "/" + hash + path;
    }

    private static byte[] hexStringToByteArray(String hex){
        if (hex.length() % 2 != 0)
            throw new IllegalArgumentException("Even-length string required");
        byte[] res = new byte[hex.length() / 2];
        for (int i = 0; i < res.length; i++) {
            res[i]=(byte)((Character.digit(hex.charAt(i * 2), 16) << 4) | (Character.digit(hex.charAt(i * 2 + 1), 16)));
        }
        return res;
    }
}

File: examples/signature.clj
(ns imgproxy-test
  (:require
   [clojure.test :refer [deftest is]])
  (:import
   (java.util Base64)
   (javax.crypto Mac)
   (javax.crypto.spec SecretKeySpec)))

(defn hex-string-to-byte-array
  [hex]
  (let [res (byte-array (/ (count hex) 2))]
    (dotimes [i (count res)]
      (aset res
            i
            (unchecked-byte (bit-or (bit-shift-left (Character/digit (nth hex (* i 2)) 16) 4)
                                    (Character/digit (nth hex (+ (* i 2) 1)) 16)))))
    res))

(defn sign-path
  [key salt path]
  (let [mac (doto (Mac/getInstance "HmacSHA256") (.init (SecretKeySpec. key "HmacSHA256")))
        _ (.update mac salt)
        hash (.doFinal mac (.getBytes path "UTF-8"))
        encoded-hash (.. (Base64/getUrlEncoder) withoutPadding (encodeToString hash))]
    (str "/" encoded-hash path)))

(deftest test-with-hmac-base64-img-proxy-test
  (is
   (=
    "/m3k5QADfcKPDj-SDI2AIogZbC3FlAXszuwhtWXYqavc/rs:fit:300:300/plain/http://img.example.com/pretty/image.jpg"
    (sign-path (hex-string-to-byte-array "943b421c9eb07c830af81030552c86009268de4e532ba2ee2eab8247c6da0881")
               (hex-string-to-byte-array "520f986b998545b4785e0defbc4f3c1203f22de2374a3d53cb7a7fe9fea309c5")
               "/rs:fit:300:300/plain/http://img.example.com/pretty/image.jpg"))))
